### PR TITLE
AWS KMS keymanager issues

### PR DESCRIPTION
**Pull request by kunzimariano**
This PR attempts to address the 2 first points from #2302, by removing unsupported characters from key ids and adding proper error handling when retrieving existing keys.

### CODE REVIEW

**Comment by azdagron** [comment]
Hey @kunzimariano, thanks for addressing these issues.

While I totally support limiting the character set used in key manager key IDs and think we'll want to come up with a safe way to move towards that goal, I'm concerned about changing those key ids now. There are two consequences: (1) it will force an ACME renewal, and (2) the old key will be left dangling. I'm not too concerned about the first. The second point  however is problematic. It is not good security policy to leave old private keys around and our key manager interface currently has no means by which private keys can be cleaned up. Updating the interface will take some time. Until then, we should probably keep the key IDs as-is.

An alternative approach would be to have the awskms key manager do some sort of encoding or mapping of the key IDs into an acceptable character set. There are probably a few solutions here.

**Comment by kunzimariano** [comment]
Thank you @azdagron!

I believe it makes a lot of sense. I considered at some point to encode the set of characters into a subset of valid ones. I will look into that or mapping as you suggested. 

Thanks again!

**Comment by kunzimariano** [comment]
Hey @azdagron 👋🏽 This new implementation makes use of a base64 encoding, except for 2 characters. Although not pretty, It also supports upgrading from a Server that was using keys without encoding, without forcing a rotation. It does not support downgrading though. 

If this approach is reasonable I'm happy to add some extra tests to cover the new scenarios. The docs might need some explanation too.

Thanks!

File: pkg/server/plugin/keymanager/awskms/awskms.go [hunk 1]
```diff
@@ -853,3 +882,23 @@ func makeFingerprint(pkixData []byte) string {
 	s := sha256.Sum256(pkixData)
 	return hex.EncodeToString(s[:])
 }
+
+// Encode into a slightly different version of base64.
+// The end result is limited to characters supported by KMS alias [a-zA-Z0-9/_-]
+func encodeKeyID(keyID string) string {
+	encoded := base64.StdEncoding.EncodeToString([]byte(keyID))
```
Comment (azdagron) [inline pkg/server/plugin/keymanager/awskms/awskms.go:889]: Isn't this encode + replacement the same as using base64.URLEncoding?

**Review by azdagron — COMMENTED** [review]
Thanks for addressing this @kunzimariano. I'm pretty concerned about the break in downgrade compatability. I wonder if we can maintain forward and backward compatability by leveraging the fact that existing key IDs have never used an underscore. We could do a mapping where an underscore becomes an escape character for a percent-encoding style encoding. Existing keys would map the same way they have today. Problematic keys that haven't worked (like the ACME one) would get encoded.

A few examples:
`x509-CA-A` ===> `x509-CA-A` (produce the same key id)
`bundle-acme-foo.bar+rsa` ===> `bundle-acme-foo_2ebar_2brsa` (. and + get mapped the `_<asciihex>`

This has the added benefit of mostly leaving the IDs intact, making them more or less identifiable for someone looking at the key IDs in the AWS console.

Thoughts?

**Comment by azdagron** [comment]
@kunzimariano @marcosy , are there parts of this PR that aren't handled by #2390? If so, is the plan to open up a separate PR for those bits (so #2390 can just focus on the ID encoding)?

**Comment by marcosy** [comment]
@azdagron, #2390 focuses only on encoding.

This PR focuses only on encoding too. I was told by @kunzimariano that the second point of #2302 is already fixed on the main branch.

**Comment by azdagron** [comment]
Closing in lieu of #2390.

**Pull request closed without merging**